jgrapht-guava/src/main/java/org/jgrapht/graph/WeightedGraphImmutableNetworkAdapter.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * (C) Copyright 2018-2018, by Dimitrios Michail and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.graph;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jgrapht.EdgeFactory;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphType;
+import org.jgrapht.util.TypeUtil;
+
+import com.google.common.graph.Graphs;
+import com.google.common.graph.ImmutableNetwork;
+import com.google.common.graph.NetworkBuilder;
+
+/**
+ * A weighted graph adapter class using Guava's {@link ImmutableNetwork}.
```

Comment:
(And I just have to say, the guava interface split seems like they were intentionally trying to make it hard to write universally applicable algorithms...e.g. how would you write the equivalent of a ClosestFirstIterator?  The predecessor/successor functions aren't helpful since they don't provide any access to the edge information.  Well, I guess once we have a complete set of adapters, that will be the answer.)

Reply:
The `ValueGraph` does not have any class for the edges, the first generic parameter is the type of the vertices and the second generic parameter is the type of the weight of the edges. 

Since our weight is `double` and not a generic parameter, this forces us to map to a graph with a custom edge class which contains the weight as a field. Ideas?

Reply:
@jsichi My understanding with the Guava `common.graph` interface split (as well as my personal experience with using both Guava and JGraphT) is that it makes things easier for users (rather than implementors or algorithm writers, necessarily), because e.g. if one doesn't need to associate info with edges, then they can just use `Graph` which has a by comparison a simpler API than `ValueGraph` and `Network`.

But you make an interesting point about `ClosestFirstIterator`. How _would_ one implement it for all 3 Guava types? :thinking:

Reply:
@d-michail How about introducing an adapter class that turns just a `ValueGraph<V, Double>` into a JGrapht `Graph`?

Alternatively, maybe we could introduce an adapter class that turns `ValueGraph<V, E>` into a JGraphT `Graph`, but the user has to pass in something like Guava's [`Converter`](https://static.javadoc.io/com.google.guava/guava/24.0-jre/com/google/common/base/Converter.html) to the constructor, where `converter.convert()` would turn a given edge of type `E` into a weight of type `Double`, and `converter.reverse().convert()` would do the opposite?

Reply:
Since @d-michail removed the weighted adapter in favor of AsWeightedGraph for now, I'll restart this conversation at top level as notes for a followup pull request.

Reply:
@jsichi Oops, I'd forgotten that @d-michail wrote that comment! Indeed, let's restart the conversation at top level. :)